### ISSUE TITLE
Update @wordpress/data to the latest version

### ISSUE THREAD

**Issue by felixarntz** [google/site-kit-wp#1769]
## Feature Description

While updating dependencies via #1356, it became apparent that there were some issues with `@wordpress/data`, which has prevented us from updating it to its latest version as well as preventing us from updating to the latest versions of `react` and `react-dom`.

Currently we're using `react` and `react-dom` 16.12.0 and `@wordpress/data` 4.12.0.

When updating `@wordpress/data` to a version newer than 4.12.0, we run into an error whenever `useSelect` is used.

When updating _only_ `react` and `react-dom` to a version newer than 16.12.0, the following error is thrown when opening a Site Kit module's settings view:
```
Warning: Cannot update a component (`%s`) while rendering a different component (`%s`). To locate the bad setState() call inside `%s`, follow the stack trace as described in https://fb.me/setstate-in-render%s PermissionsModal AdBlockerWarning AdBlockerWarning 
        in AdBlockerWarning
        in Unknown
        in InnerComponent (created by WithFilters(ModuleSettingsWarning))
        in WithFilters(ModuleSettingsWarning)
        in div
        in SetupModule
        in SettingsModule
        in div
        in div
        in div
        in div
        in Layout
        in div
        in SettingsModules
        in div
        in div
        in div
        in SettingsApp
        in RestoreSnapshots
        in ErrorHandler
        in Root
```

Looking into the latter, it seems this problem is also somehow caused by the `useSelect` hook implemented by `@wordpress/data`. Let's dive deeper into what the problem is here. (Note that WordPress packages currently rely on an older React version 16.9.0.)

* [`useSelect` which we're using](https://github.com/WordPress/gutenberg/blob/b6ff6e99c76afeb0e8e1b784f8ea5ae48cd45f10/packages/data/src/components/use-select/index.js)
* [latest `useSelect`](https://github.com/WordPress/gutenberg/blob/master/packages/data/src/components/use-select/index.js)

Potentially we will end up blocked by `@wordpress/data` fixing the underlying issue - let's file an upstream issue / pull-request about the problem if we can figure out what it is.

---------------

_Do not alter or remove anything below. The following sections will be managed by moderators only._

## Acceptance criteria

* `@wordpress/data` should be updated to ~~its latest version (currently 4.27.3)~~ `4.23.0` (as a first step towards later versions).
    * `react` and `react-dom` may be updated but only as necessary
* No JavaScript errors and no new warnings should be triggered when visiting any Site Kit UI.

## Implementation Brief

Please note: This IB has been scoped to updating to `@wordpress/data` version 4.23.0, as that is the version that introduces the change that breaks the majority of our tests (see [this comment](https://github.com/google/site-kit-wp/issues/1769#issuecomment-1221335331) below). Subsequent issues should be raised, one for updating to version 4.27.3 (the latest version that supports React 16), and another to then investigate updating to a more recent version which supports React 17 (currently, the latest `@wordpress/data` version is 7.6.0).

- Update to `@wordpress/data` version 4.23.0.
- Ensure the plugin continues to work as usual.
- Create a followup issue to upgrade to version 4.27.3.
- Create a subsequent followup issue to investigate updating to the latest version.

### Test Coverage 

- Fix all failing tests.
- The most effective way of fixing the tests takes its initial cue from the [previous PR](https://github.com/google/site-kit-wp/pull/3277) for the issue which is now a bit out of date. That is, to enable Jest's real timers for all tests in a global `beforeAll`. This reduces the initial failing test count from 813 to 213 at the time of writing. However when it comes to fixing the remaining tests we can also use the approach that the WordPress devs use (again see the [comment](https://github.com/google/site-kit-wp/issues/1769#issuecomment-1221335331) below), switching to fake timers and calling `jest.runAllTimers()` as needed. Additionally the techniques seen in the previous PR are still useful.
- A [new PR](https://github.com/google/site-kit-wp/pull/6203) to kick this off has been created which fixes a number of these remaining tests (55, to be precise) in a few different ways. It's working in the browser, although there are a couple of failing E2E tests to take a look at. The VRT tests are currently failing but that's a general [issue](https://github.com/google/site-kit-wp/issues/6185), although for safety's sake we should wait for #6185 to be merged so we can make a proper comparison. Storybook itself appears to be OK.
- The above mentioned [PR](https://github.com/google/site-kit-wp/pull/6203) should be picked up and finished off.

<details>

<summary>
Note: The list of remaining failing test suites is as follows:
</summary>

This is correct at the time of writing, to recreate the list run this command:

```sh
> npm run test:js 2>&1 | grep FAIL | sort | uniq
```

```
assets/js/googlesitekit/datastore/user/authentication.test.js (11.239s)
assets/js/googlesitekit/datastore/user/dismissed-items.test.js
assets/js/googlesitekit/datastore/user/feature-tours.test.js (11.267s)
assets/js/googlesitekit/datastore/user/permissions.test.js
assets/js/googlesitekit/datastore/user/surveys.test.js
assets/js/googlesitekit/datastore/user/user-info.test.js (11.282s)
assets/js/googlesitekit/modules/datastore/modules.test.js (12.046s)
assets/js/googlesitekit/modules/datastore/sharing-settings.test.js
assets/js/hooks/useChecks.test.js
assets/js/hooks/useCompleteModuleActivationCallback.test.js
assets/js/hooks/useDebouncedState.test.js
assets/js/modules/adsense/components/module/ModuleOverviewWidget/Footer.test.js
assets/js/modules/adsense/datastore/accounts.test.js
assets/js/modules/adsense/datastore/adblocker.test.js
assets/js/modules/adsense/datastore/alerts.test.js
assets/js/modules/adsense/datastore/clients.test.js
assets/js/modules/adsense/datastore/report.test.js
assets/js/modules/adsense/datastore/sites.test.js
assets/js/modules/analytics-4/datastore/webdatastreams.test.js (11.164s)
assets/js/modules/analytics/components/common/AccountSelect.test.js
assets/js/modules/analytics/components/common/PropertySelectIncludingGA4.test.js
assets/js/modules/analytics/components/module/ModulePopularPagesWidget/Footer.test.js
assets/js/modules/analytics/components/settings/SettingsEdit.test.js
assets/js/modules/analytics/datastore/accounts.test.js (11.474s)
assets/js/modules/analytics/datastore/adsense.test.js
assets/js/modules/analytics/datastore/goals.test.js
assets/js/modules/analytics/datastore/profiles.test.js
assets/js/modules/analytics/datastore/properties.test.js
assets/js/modules/analytics/datastore/report.test.js (11.053s)
assets/js/modules/analytics/datastore/setup-flow.test.js
assets/js/modules/idea-hub/components/dashboard/DashboardIdeasWidget/index.test.js (16.465s)
assets/js/modules/pagespeed-insights/datastore/report.test.js
assets/js/modules/search-console/components/dashboard/SearchFunnelWidget/Footer.test.js
assets/js/modules/search-console/components/dashboard/SearchFunnelWidget/index.test.js
assets/js/modules/search-console/components/dashboard/SearchFunnelWidget/Overview.test.js
assets/js/modules/search-console/datastore/report.test.js (10.995s)
assets/js/modules/tagmanager/components/settings/SettingsEdit.test.js (14.372s)
assets/js/modules/tagmanager/components/setup/SetupMain.test.js (14.197s)
assets/js/modules/tagmanager/datastore/accounts.test.js (11.965s)
assets/js/modules/tagmanager/datastore/containers.test.js
assets/js/modules/tagmanager/datastore/versions.test.js
assets/js/modules/thank-with-google/datastore/publications.test.js
```

</details>

## QA Brief

- This needs a really thorough regression test of the entirety of Site Kit - every module and feature.
- Focus on the behavioural aspects, this is an upgrade of the JS data library that underpins the UI, so we need to be sure the plugin still behaves as expected. It's unrelated to the styling, though, so there's no need to dig into that too deeply, as long as it all looks as expected it can be assumed the styles are unaffected.
- This change also includes a minor React version bump, 16.12.0 -> 16.14.0. All Site Kit UI should be checked for React errors but also non-fatal React errors/warnings in the console.

## Changelog entry

* Upgrade `@wordpress/data` to 4.23.0, `react` and `react-dom` to 16.14.0.

**Comment by felixarntz** [comment]
@tofumatt This is still blocked by #1761 though right?

**Comment by tofumatt** [comment]
These are the warnings I see in the console, but they aren't errors and don't impact tests/usability of the plugin:

<img width="1389" alt="Screenshot 2020-08-18 at 12 56 11" src="https://user-images.githubusercontent.com/90871/90510066-37535f00-e152-11ea-9e18-d175957d5a29.png">

That said, we could mark this as blocked by that issue so these warnings don't appear in the plugin. They do only appear in development mode, so the warnings won't appear in the production version of the plugin (here's the same site running the version created by `npm run build`:

<img width="1390" alt="Screenshot 2020-08-18 at 12 59 06" src="https://user-images.githubusercontent.com/90871/90510294-9fa24080-e152-11ea-9e92-f47c0b300e83.png">

I wouldn't consider it a blocker because those notices only appear in development.

**Comment by felixarntz** [comment]
IB ✅

**Comment by tofumatt** [comment]
Further testing of this shows `useInstanceId` is causing hooks errors even without `@wordpress/data` or React upgraded.

**Comment by tofumatt** [comment]
Moving this back to the execution backlog with a much higher estimate, as this issue has totally eluded me.

Upgrading `@wordpress/data` to `4.13` from `4.12` breaks nearly every test, but strangely it seems to only be in the tests—largely Jest tests. The diff is here: https://diff.intrinsic.com/@wordpress/data/4.12.0/4.13.0#file-src__components__use-dispatch__use-dispatch-with-map.js and aside from a move to `useMemoOne` there isn't a lot there that's changed.

We should plan to upgrade these critical dependencies, but it will take a lot more time.

------

That said, there's a further discussion to be had about our usage of `@wordpress/data`. It's a system largely only used by Gutenberg and seemingly not as a third-party dependency by many other projects directly. The changelogs are not reliable/up-to-date, and we are possibly one of the larger non-WordPress Core projects using it. When we encounter issues like this, it's tough to find others with similar problems. Something to consider, but that's about all for now 😄

**Comment by felixarntz** [comment]
Moving this back to IB because the current IB was against the older requirements when we weren't aware of the issues around upgrading `@wordpress/data`.

**Comment by felixarntz** [comment]
Moving to _Stalled_ - we can revisit this once the refactoring is complete and we have some time to dive into this.

**Comment by aaemnnosttv** [comment]
@felixarntz @tofumatt please assign back to me when ready to proceed.

**Comment by tofumatt** [comment]
IB ✅

**Comment by eugene-manuilov** [comment]
@aaemnnosttv or @tofumatt can anyone of you help with this ticket? Unfortunately, I can't solve all js test issues. :cry:

**Comment by eclarke1** [comment]
@felixarntz @aaemnnosttv just a friendly question about this issue and whether it is still needed, or could we define next steps? This one has been rolling over from sprint to sprint since September so wondered if there's a decision we need to make here? cc @FlicHollis

**Comment by aaemnnosttv** [comment]
It is still needed, but has lagged a bit in priority. The scale of the changes needed and quirks to work through along with familiarity with the datastore makes it a bit limited as to who can pick it up. I'd love to pick this back up again but haven't had the time.

**Comment by FlicHollis** [comment]
Hi @aaemnnosttv Moving this back to EB as this has  not been added to Sprint 68

**Comment by techanvil** [comment]
> @felixarntz I spent some time looking into this today related to another issue and was able to get it working in a way that seems normal in the browser with no errors. It required upgrading React to 16.14 which is technically out of scope as defined but seems to be necessary. Despite things working normally in the browser, there were quite a few JS test failures but these all seemed to be related to tests around resolvers. Basically, there seem to be some things that were happening synchronously before that aren't and we need to wait for a resolver to finish before asserting something else like `toHaveErrored` or `toHaveFetched` so I think the rest of the work should be pretty straightforward.
> 
> Let's move this forward as I don't think there is as much left to do here as we may have thought.

I wanted to get a deeper understanding of the cause of these test failures, so dug into it a bit. It looks the change that blew up most of the tests is the introduction of a resolver cache and related call to `setTimeout`, introduced in `@wordpress/data@4.23.0`. I found that updating as far as `4.22.3`, all JS tests passed, then bumping to `4.23.0` hundreds of tests started failing.

Here is the diff `@wordpress/data@4.22.3` -> `4.23.0`. It's interesting to see the approach taken to fixing the tests - this can give an extra angle to our own fixes.

<details>

```diff
diff --git a/packages/data/README.md b/packages/data/README.md
index fc904d64e3..4fa08b44bb 100644
--- a/packages/data/README.md
+++ b/packages/data/README.md
@@ -796,4 +796,8 @@ _Returns_
 
 <!-- END TOKEN(Autogenerated API docs) -->
 
+## Going further
+
+-   [What is WordPress Data?](https://unfoldingneurons.com/2020/what-is-wordpress-data/)
+
 <br/><br/><p align="center"><img src="https://s.w.org/style/images/codeispoetry.png?1" alt="Code is Poetry." /></p>
diff --git a/packages/data/package.json b/packages/data/package.json
index 958e9c4bba..1f238eac7e 100644
--- a/packages/data/package.json
+++ b/packages/data/package.json
@@ -1,6 +1,6 @@
 {
 	"name": "@wordpress/data",
-	"version": "4.22.3",
+	"version": "4.23.0",
 	"description": "Data module for WordPress.",
 	"author": "The WordPress Contributors",
 	"license": "GPL-2.0-or-later",
@@ -33,7 +33,7 @@
 		"@wordpress/redux-routine": "file:../redux-routine",
 		"equivalent-key-map": "^0.2.2",
 		"is-promise": "^4.0.0",
-		"lodash": "^4.17.15",
+		"lodash": "^4.17.19",
 		"memize": "^1.1.0",
 		"redux": "^4.0.0",
 		"turbo-combine-reducers": "^1.0.2",
diff --git a/packages/data/src/namespace-store/index.js b/packages/data/src/namespace-store/index.js
index a3bcf5402b..952c620acf 100644
--- a/packages/data/src/namespace-store/index.js
+++ b/packages/data/src/namespace-store/index.js
@@ -4,6 +4,7 @@
 import { createStore, applyMiddleware } from 'redux';
 import { flowRight, get, mapValues } from 'lodash';
 import combineReducers from 'turbo-combine-reducers';
+import EquivalentKeyMap from 'equivalent-key-map';
 
 /**
  * WordPress dependencies
@@ -19,6 +20,34 @@ import metadataReducer from './metadata/reducer';
 import * as metadataSelectors from './metadata/selectors';
 import * as metadataActions from './metadata/actions';
 
+/**
+ * Create a cache to track whether resolvers started running or not.
+ *
+ * @return {Object} Resolvers Cache.
+ */
+function createResolversCache() {
+	const cache = {};
+	return {
+		isRunning( selectorName, args ) {
+			return cache[ selectorName ] && cache[ selectorName ].get( args );
+		},
+
+		clear( selectorName, args ) {
+			if ( cache[ selectorName ] ) {
+				cache[ selectorName ].delete( args );
+			}
+		},
+
+		markAsRunning( selectorName, args ) {
+			if ( ! cache[ selectorName ] ) {
+				cache[ selectorName ] = new EquivalentKeyMap();
+			}
+
+			cache[ selectorName ].set( args, true );
+		},
+	};
+}
+
 /**
  * @typedef {WPDataRegistry} WPDataRegistry
  */
@@ -37,6 +66,7 @@ import * as metadataActions from './metadata/actions';
 export default function createNamespace( key, options, registry ) {
 	const reducer = options.reducer;
 	const store = createReduxStore( key, options, registry );
+	const resolversCache = createResolversCache();
 
 	let resolvers;
 	const actions = mapActions(
@@ -64,7 +94,12 @@ export default function createNamespace( key, options, registry ) {
 		store
 	);
 	if ( options.resolvers ) {
-		const result = mapResolvers( options.resolvers, selectors, store );
+		const result = mapResolvers(
+			options.resolvers,
+			selectors,
+			store,
+			resolversCache
+		);
 		resolvers = result.resolvers;
 		selectors = result.selectors;
 	}
@@ -166,7 +201,6 @@ function createReduxStore( key, options, registry ) {
  *                           public facing API. Selectors will get passed the
  *                           state as first argument.
  * @param {Object} store     The store to which the selectors should be mapped.
- *
  * @return {Object} Selectors mapped to the provided store.
  */
 function mapSelectors( selectors, store ) {
@@ -215,12 +249,13 @@ function mapActions( actions, store ) {
  * Resolvers are side effects invoked once per argument set of a given selector call,
  * used in ensuring that the data needs for the selector are satisfied.
  *
- * @param {Object} resolvers   Resolvers to register.
- * @param {Object} selectors   The current selectors to be modified.
- * @param {Object} store       The redux store to which the resolvers should be mapped.
- * @return {Object}            An object containing updated selectors and resolvers.
+ * @param {Object} resolvers      Resolvers to register.
+ * @param {Object} selectors      The current selectors to be modified.
+ * @param {Object} store          The redux store to which the resolvers should be mapped.
+ * @param {Object} queue          Resolvers async queue.
+ * @param {Object} resolversCache Resolvers Cache.
  */
-function mapResolvers( resolvers, selectors, store ) {
+function mapResolvers( resolvers, selectors, store, resolversCache ) {
 	const mappedResolvers = mapValues( resolvers, ( resolver ) => {
 		const { fulfill: resolverFulfill = resolver } = resolver;
 		return { ...resolver, fulfill: resolverFulfill };
@@ -237,13 +272,15 @@ function mapResolvers( resolvers, selectors, store ) {
 			async function fulfillSelector() {
 				const state = store.getState();
 				if (
-					typeof resolver.isFulfilled === 'function' &&
-					resolver.isFulfilled( state, ...args )
+					resolversCache.isRunning( selectorName, args ) ||
+					( typeof resolver.isFulfilled === 'function' &&
+						resolver.isFulfilled( state, ...args ) )
 				) {
 					return;
 				}
 
 				const { metadata } = store.__unstableOriginalGetState();
+
 				if (
 					metadataSelectors.hasStartedResolution(
 						metadata,
@@ -254,18 +291,23 @@ function mapResolvers( resolvers, selectors, store ) {
 					return;
 				}
 
-				store.dispatch(
-					metadataActions.startResolution( selectorName, args )
-				);
-				await fulfillResolver(
-					store,
-					mappedResolvers,
-					selectorName,
-					...args
-				);
-				store.dispatch(
-					metadataActions.finishResolution( selectorName, args )
-				);
+				resolversCache.markAsRunning( selectorName, args );
+
+				setTimeout( async () => {
+					resolversCache.clear( selectorName, args );
+					store.dispatch(
+						metadataActions.startResolution( selectorName, args )
+					);
+					await fulfillResolver(
+						store,
+						mappedResolvers,
+						selectorName,
+						...args
+					);
+					store.dispatch(
+						metadataActions.finishResolution( selectorName, args )
+					);
+				} );
 			}
 
 			fulfillSelector( ...args );
diff --git a/packages/data/src/namespace-store/test/index.js b/packages/data/src/namespace-store/test/index.js
index 4d0ca7dc74..0198f59720 100644
--- a/packages/data/src/namespace-store/test/index.js
+++ b/packages/data/src/namespace-store/test/index.js
@@ -4,6 +4,8 @@
 import { createRegistry } from '../../registry';
 import { createRegistryControl } from '../../factory';
 
+jest.useFakeTimers();
+
 describe( 'controls', () => {
 	let registry;
 
@@ -80,16 +82,14 @@ describe( 'controls', () => {
 					.select( 'store' )
 					.hasFinishedResolution( 'getItems' );
 				if ( isFinished ) {
-					expect( registry.select( 'store' ).getItems() ).toEqual( [
-						1,
-						2,
-						3,
-					] );
+					const items = registry.select( 'store' ).getItems();
+					expect( items ).toEqual( [ 1, 2, 3 ] );
 				}
 				resolve();
 			} );
 
 			registry.select( 'store' ).getItems();
+			jest.runAllTimers();
 		} );
 	} );
 	describe( 'selectors have expected value for the `hasResolver` property', () => {
diff --git a/packages/data/src/test/registry.js b/packages/data/src/test/registry.js
index 879050388a..1d78e8dae1 100644
--- a/packages/data/src/test/registry.js
+++ b/packages/data/src/test/registry.js
@@ -9,6 +9,8 @@ import { castArray, mapValues } from 'lodash';
 import { createRegistry } from '../registry';
 import { createRegistrySelector } from '../factory';
 
+jest.useFakeTimers();
+
 describe( 'createRegistry', () => {
 	let registry;
 
@@ -261,11 +263,14 @@ describe( 'createRegistry', () => {
 			} );
 
 			const value = registry.select( 'demo' ).getValue( 'arg1', 'arg2' );
+			jest.runAllTimers();
 			expect( value ).toBe( 'OK' );
 			expect( resolver ).toHaveBeenCalledWith( 'arg1', 'arg2' );
 			registry.select( 'demo' ).getValue( 'arg1', 'arg2' );
+			jest.runAllTimers();
 			expect( resolver ).toHaveBeenCalledTimes( 1 );
 			registry.select( 'demo' ).getValue( 'arg3', 'arg4' );
+			jest.runAllTimers();
 			expect( resolver ).toHaveBeenCalledTimes( 2 );
 		} );
 
@@ -282,6 +287,7 @@ describe( 'createRegistry', () => {
 			} );
 
 			const value = registry.select( 'demo' ).getValue( 'arg1', 'arg2' );
+			jest.runAllTimers();
 			expect( value ).toBe( 'OK' );
 		} );
 
@@ -317,21 +323,29 @@ describe( 'createRegistry', () => {
 
 			store.dispatch( { type: 'SET_PAGE', page: 4, result: [] } );
 			registry.select( 'demo' ).getPage( 1 );
+			jest.runAllTimers();
 			registry.select( 'demo' ).getPage( 2 );
+			jest.runAllTimers();
 
 			expect( fulfill ).toHaveBeenCalledTimes( 2 );
 
 			registry.select( 'demo' ).getPage( 1 );
+			jest.runAllTimers();
 			registry.select( 'demo' ).getPage( 2 );
+			jest.runAllTimers();
 			registry.select( 'demo' ).getPage( 3, {} );
+			jest.runAllTimers();
 
 			// Expected: First and second page fulfillments already triggered, so
 			// should only be one more than previous assertion set.
 			expect( fulfill ).toHaveBeenCalledTimes( 3 );
 
 			registry.select( 'demo' ).getPage( 1 );
+			jest.runAllTimers();
 			registry.select( 'demo' ).getPage( 2 );
+			jest.runAllTimers();
 			registry.select( 'demo' ).getPage( 3, {} );
+			jest.runAllTimers();
 			registry.select( 'demo' ).getPage( 4 );
 
 			// Expected:
@@ -366,6 +380,7 @@ describe( 'createRegistry', () => {
 			] );
 
 			registry.select( 'demo' ).getValue();
+			jest.runAllTimers();
 
 			return promise;
 		} );
@@ -392,6 +407,7 @@ describe( 'createRegistry', () => {
 			] );
 
 			registry.select( 'demo' ).getValue();
+			jest.runAllTimers();
 
 			return promise;
 		} );
@@ -416,6 +432,7 @@ describe( 'createRegistry', () => {
 			shouldThrow = true;
 
 			registry.select( 'demo' ).getValue();
+			jest.runAllTimers();
 
 			return new Promise( ( resolve ) => process.nextTick( resolve ) );
 		} );
@@ -440,7 +457,9 @@ describe( 'createRegistry', () => {
 			);
 
 			registry.select( 'demo' ).getValue();
+			jest.runAllTimers();
 			registry.select( 'demo' ).getValue();
+			jest.runAllTimers();
 
 			return promise;
 		} );
@@ -471,6 +490,7 @@ describe( 'createRegistry', () => {
 				() => registry.select( 'demo' ).getValue() === 'OK'
 			);
 			registry.select( 'demo' ).getValue(); // Triggers resolver switches to OK
+			jest.runAllTimers();
 			await promise;
 
 			// Invalidate the cache
@@ -480,6 +500,7 @@ describe( 'createRegistry', () => {
 				() => registry.select( 'demo' ).getValue() === 'NOTOK'
 			);
 			registry.select( 'demo' ).getValue(); // Triggers the resolver again and switch to NOTOK
+			jest.runAllTimers();
 			await promise;
 		} );
 	} );
```

</details>

---

In case anyone finds it useful (I might myself, if I revisit this), here's a little script that loops over the `@wordpress/data` versions, installing each version in turn and running the JS integration test suite, until the test suite fails. It assumes the Gutenberg repo is cloned to `$WordPress_gutenberg`, and would need tweaking if `git tag` output has moved on the meantime.

```sh
for VERSION in $( \
  cd $WordPress_gutenberg >/dev/null && \
  git tag | grep /data | egrep '@[456].*' | grep -v '4\.2\.' | egrep '[456].+' | tail -57 \
); do
  npm i $VERSION && git add . && git commit --no-verify -m "Bump to $VERSION." && \
  npm run test:js || break
done
```

**Comment by aaemnnosttv** [comment]
@techanvil thanks for digging into this, the added context is really useful!

Looking at the changes to the tests in Gutenberg, I wonder if we could use the same approach to call `jest.runAllTimers()` in  an `afterEach` rather than using real timers as was done in the original PR 🤔 https://github.com/google/site-kit-wp/pull/3277/files#diff-8562ee22152539b10725dc937f9daa64323a8897a0e28bfb640a7c346f4d9f72R35

It would be great to finally see this one completed, and this definitely helps to clarify the path forward 👏 Great work!

**Comment by techanvil** [comment]
Thanks @aaemnnosttv, I have taken another look at this, keen to move it along. I tried your suggestion of running `jest.runAllTimers()` in an `afterEach`, but it wasn't very successful, I had more luck continuing with the real timers as you had done, but bringing the fake timers in to fix individual tests.

I've updated the IB, I think we should be in a good position to make some progress with this one. Please see what you think.

**Comment by FlicHollis** [comment]
Morning @techanvil please could you add an estimate to this ticket? Thanks so much!

**Comment by FlicHollis** [comment]
Thanks so much @techanvil and good to know!

**Comment by aaemnnosttv** [comment]
> * [...] we should first upgrade to the latest `@wordpress/data` version that supports React 16, which is version 4.27.3.
> * Therefore this issue should be repurposed to upgrade to `@wordpress/data` version 4.23.0. A followup issue should be raised to upgrade to 4.27.3, and another to then look beyond this to a more recent version, where we can also consider upgrading our major React version.

@techanvil the IB seems to conflict with itself because it suggests upgrading to 4.27.3 and 4.23.0 in separate places. The AC also mentions the former so if it makes more sense to update that in the process, that's fine with me but let's make sure everything reads correctly :)

Aside from that, I'm a bit skeptical that this can be done as a 15 given the history here. How about a 30? 😄

**Comment by techanvil** [comment]
Hi @aaemnnosttv, thanks for the feedback. Sorry for the confusion in the IB, I've updated it to be more clear. I don't think we should try to update beyond 4.23.0 in this issue as there are so many tests to update for the bump to 4.23.0, it's best to keep things as simple as possible and not cloud things by bringing later versions into the mix. Once we've got 4.23.0 addressed, the next update should be a lot simpler.

The reason I estimated 15 is because I got on a bit of a roll with fixing tests and managed to get ~50 working in the space of a couple hours or so. And we do have the new `runAllTimers` tool in the kit for this one which helps to simplify things a bit. However, I fully agree that it could easily end up taking much longer, with all those tests to fix, there are a lot of unknowns. Entirely happy to be a bit more cautious and bump it to a 30! We are supposed to split things up above 19 points but it's not really possible with this one - unless we were to use a feature branch, although I don't see that as being particularly useful here.

**Comment by aaemnnosttv** [comment]
Sounds good, thanks for clarifying @techanvil !

IB ✅ (part deux 🚀)

**Comment by aaemnnosttv** [comment]
@techanvil would you please open the next you've issues identified here?

**Comment by techanvil** [comment]
Thanks for calling that out @aaemnnosttv. I've created followups #6356 and #6357.

**Comment by aaemnnosttv** [comment]
I noticed a console error (React warning) when changing the date range on the dashboard, but this appears to be coming from `react-google-charts`. It does not seem to happen in the latest release but this warning will only appear in development builds and is probably related to the update to React here. It's not something to block the issue or release over so we can look into addressing this going forward, as it is also [a known issue](https://github.com/rakannimer/react-google-charts/issues?q=is%3Aissue+is%3Aopen+sort%3Aupdated-desc+unmounted) on the chart package repo.

<details>
<summary>Screenshot</summary>
![image](https://user-images.githubusercontent.com/1621608/212181689-07d842be-2e5e-4a71-a9b6-dc2d8d5be31e.png)
</details>

Happy to finally close this one out after more than 2 years in the making 🎉

**Comment by wpdarren** [comment]
> It's not something to block the issue or release over so we can look into addressing this going forward

@aaemnnosttv is it worth creating a ticket for this?
